File: python/trie.py
class TrieNode(object):
      def __init__(self):
        self.end = False
        self.trieSet = {}

class Trie(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()


    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        if self.search(word):
            return

        curr = self.root
        for c in word:
            if c not in curr.trieSet:
                curr.trieSet[c] = TrieNode()
            curr = curr.trieSet[c]

        if curr:
            curr.end = True


    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        return self.searchImpl(word, False)

    def searchImpl(self, word, startsWith):

        if not word:
            return False

        curr = self.root
        for c in word:
            if c in curr.trieSet:
                curr = curr.trieSet[c]
            else:
                return False

        if startsWith:
            return True
        else:
            return curr and curr.end


    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        return self.searchImpl(prefix, True)

if __name__ == '__main__':
        tree = Trie()
        tree.insert("");
        assert(tree.search("") == False);

        tree.insert("a");
        assert(tree.search("a") == True);

        tree.insert("ab");

        assert(tree.search("ab"));
        assert(tree.search("cd") == False);
        assert(tree.search("c") == False);

        tree.insert("cd");
        assert(tree.search("cd"));

        tree.insert("abe");
        assert(tree.search("a"));
        assert(tree.startsWith("a"));
        assert(tree.startsWith("ab"));
        assert(tree.search("ab"));
        assert(tree.search("abc") == False);

        assert(tree.search("abe"));
        assert(tree.search("c") == False);
        assert(tree.startsWith("c"));
        assert(tree.search("cd"));

        tree2 = Trie();

        tree2.insert("app");
        tree2.insert("apple");
        tree2.insert("beer");
        tree2.insert("add");
        tree2.insert("jam");
        tree2.insert("rental");
        assert(tree2.search("app"));
        assert(tree2.search("apps") == False);

File: c++/trie.cpp
#include <iostream>
#include <queue>
#include <cassert>
using namespace std;

/* g++ trie.cpp -o trie -std=c++0x */
struct TrieNode
{
    TrieNode()
    {
        for (int i = 0; i < 26 ;i++)
            trieSet[i] = nullptr;
        end = false;
    }
    TrieNode* trieSet [26];
    bool end;

};

/* Implement a trie with insert, search, and startsWith methods.

Note:
You may assume that all inputs are consist of lowercase letters a-z.
*/
class Trie
{
public:

    TrieNode* root;

    Trie()
    {
        root = new TrieNode();
        root->end = false;
    }

    /** Inserts a word into the trie. */
    void insert(string word)
    {
        if (search(word))
            return;

        if (word.size() == 0)
            return;

        TrieNode* curr = root;
        TrieNode* nextNode = nullptr;
        for (char c : word)
        {
            nextNode = curr->trieSet[getIndex(c)]; // get the next child potentially containing the current character
            if (!nextNode) // doesn't exist, so allocate
            {
                curr->trieSet[getIndex(c)] = new TrieNode();
                nextNode = curr->trieSet[getIndex(c)]; // next iteration we work on this also
            }
            curr = nextNode;
        }

        if (curr)
            curr->end = true;
    }

    /** Returns if the word is in the trie. */
    bool search(std::string word) const
    {
        return searchWrapper(word, root, false);
    }

    bool searchWrapper(std::string word, TrieNode* curr, bool prefixSearch) const
    {
       if (word.empty())
            return false;

        for (int i = 0; i < word.size(); i++)
        {
            char c = word[i];
            if (curr && curr->trieSet[getIndex(c)])
                curr = curr->trieSet[getIndex(c)];
            else
                return false;
        }
        return curr && (prefixSearch ? true : curr->end);
    }

    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix)
    {
        return searchWrapper(prefix, root, true);
    }

    int getIndex(char c) const
    {
        return (int) (c - 'a');
    }
    void print()
    {
        queue<TrieNode*> level;
        level.push(root);

        TrieNode* curr;
        while (!level.empty())
        {
            curr = level.front();
            level.pop();

            for (int i = 0; i < 26; i++)
            {
                if (curr->trieSet[i])
                {
                    std::cout << (char) (i + 'a') << ':' << (int)curr->trieSet[i]->end <<',';
                    level.push(curr->trieSet[i]);
                }
            }
            std::cout << endl;
        }
    }

};


int main()
{
    Trie tree;
    tree.insert("");
    assert(!tree.search(""));
    tree.insert("a");
    assert(tree.search("a"));

    tree.insert("ab");

    assert(tree.search("ab"));
    tree.insert("cd");
    assert(tree.search("cd"));
    tree.insert("abe");
    tree.print();
    assert(tree.search("a"));
    assert(tree.startsWith("a"));
    assert(tree.startsWith("ab"));
    assert(tree.search("ab"));
    assert(!tree.search("abc"));

    assert(tree.search("abe"));
    assert(!tree.search("c"));
    assert(tree.startsWith("c"));
    assert(tree.search("cd"));


    Trie tree2;
    tree2.insert("app");

    tree2.insert("apple");
    tree2.insert("beer");
    tree2.insert("add");
    tree2.insert("jam");
    tree2.insert("rental");
    assert(tree2.search("app"));
    assert(!tree2.search("apps"));
}
